core/capabilities/webapi/outgoing_connector_handler.go

```diff
@@ -153,6 +144,82 @@ func (c *OutgoingConnectorHandler) HandleSingleNodeRequest(ctx context.Context,
 	}
 }
 
+// AwaitConnection attempts to establish a connection to an available gateway.  It iterates through available gateways
+// using the gatewaySelector, attempting to connect to each.  The method respects the provided context, allowing for
+// cancellation or timeout.
+func (c *OutgoingConnectorHandler) AwaitConnection(ctx context.Context) (string, error) {
+	return c.awaitConnectionUntilCanceled(ctx)
+}
+
+// awaitConnectionUntilCanceled cycles through gateways via a round robin and attempts a connection.
+func (c *OutgoingConnectorHandler) awaitConnectionUntilCanceled(ctx context.Context) (string, error) {
+	selector := NewRoundRobinSelector(c.gc.GatewayIDs())
+	attempts := make(map[string]int)
+	wait := 10 * time.Millisecond
+	for {
+		select {
+		case <-ctx.Done():
+			return "", ctx.Err()
+		default:
+			gateway, err := selector.NextGateway()
+			if err != nil {
+				return "", fmt.Errorf("failed to select gateway: %w", err)
+			}
+
+			// Cycling through gateways via round robin, if we have attempted this connection before, then we have
+			// seen them all, backoff before starting over.
+			if attempts[gateway] > 0 {
+				c.lggr.Warnw("all available gateway nodes attempted without connection, backing off", "waitTime", wait)
+				attempts = make(map[string]int)
+				<-time.After(wait)
+				wait *= 2
+			}
+
+			attempts[gateway]++
+
+			c.lggr.Infow("selected gateway, awaiting connection", "selectedGateway", gateway)
+
+			if err := c.attemptGatewayConnection(ctx, gateway); err != nil {
+				c.lggr.Warnw("failed to await connection to gateway node, retrying", "selectedGateway", gateway, "error", err)
+				continue
+			}
+			return gateway, nil
+		}
+	}
+}
+
+// attemptGatewayConnection uses at most defaultAwaitTimeoutMs to connect to a gateway.
+func (c *OutgoingConnectorHandler) attemptGatewayConnection(ctx context.Context, gateway string) error {
+	et := effectiveTimeout(ctx)
+
+	c.lggr.Debugw("await connection with deadline", "timeout", et)
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, et)
+	defer cancel()
+
+	if err := c.gc.AwaitConnection(ctxWithTimeout, gateway); err != nil {
+		return fmt.Errorf("gateway connection failed: %w", err)
+	}
+	return nil
+}
+
+// effectiveTimeout returns a default duration if there is no deadline on the context, otherwise returns the time until
+// the deadline.
+func effectiveTimeout(ctx context.Context) time.Duration {
```

Comment:
@MStreet3 Is the behaviour of this the same if you just do `context.WithTimeout(ctx, defaultTimeout)`?

Whichever comes first in the context chain will fire first, so I think that should be fine

Reply:
right, parent will always cancel the derived one.  changed.